File: js/src/v1/utils/NumberValidator.ts
export class NumberValidator {
    private static regex = /^\+?(0|\d{3})\d{9}$/;

    public static validateNumbers(numbers: string[]): string[] {
        if (!numbers || numbers.length === 0) {
            console.log('Number list cannot be null or empty');
            return [];
        }

        const cleansed: string[] = [];
        for (const number of numbers) {
            if (!number || number.trim().length === 0) {
                console.log(`Number (${number}) cannot be null or empty!`);
                continue;
            }

            let cleanedNumber = number.replace(/-|\s/g, '').trim();
            if (NumberValidator.regex.test(cleanedNumber)) {
                if (cleanedNumber.startsWith('0')) {
                    cleanedNumber = '256' + cleanedNumber.substring(1);
                } else if (cleanedNumber.startsWith('+')) {
                    cleanedNumber = cleanedNumber.substring(1);
                }
                cleansed.push(cleanedNumber);
            } else {
                console.log(`Number (${number}) is not valid!`);
            }
        }
        return [...new Set(cleansed)];
    }
}

File: c#/src/v1/utils/NumberValidator.cs
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;

namespace Comms.Utils;

public static class NumberValidator
{
    private static readonly Regex Regex = new(@"^\+?(0|\d{3})\d{9}$");

    public static List<string> ValidateNumbers(List<string> numbers)
    {
        if (numbers == null || !numbers.Any())
        {
            System.Console.Error.WriteLine("Number list cannot be null or empty");
            return new List<string>();
        }

        var cleansed = new HashSet<string>();
        foreach (var number in numbers)
        {
            if (string.IsNullOrWhiteSpace(number))
            {
                System.Console.Error.WriteLine($"Number ({number}) cannot be null or empty!");
                continue;
            }

            var trimmedNumber = number.Trim().Replace("-", "").Replace(" ", "");
            if (Regex.IsMatch(trimmedNumber))
            {
                if (trimmedNumber.StartsWith("0"))
                {
                    trimmedNumber = "256" + trimmedNumber.Substring(1);
                }
                else if (trimmedNumber.StartsWith("+"))
                {
                    trimmedNumber = trimmedNumber.Substring(1);
                }
                cleansed.Add(trimmedNumber);
            }
            else
            {
                System.Console.Error.WriteLine($"Number ({number}) is not valid!");
            }
        }
        return cleansed.ToList();
    }
}
